api/server/app.cc

```diff
@@ -456,34 +458,52 @@ static const char *SrcCgroup() {
     return NULL;
 }
 
+int setCgroups(std::string file1, std::string file2) {
+    std::ifstream ifs(file1);
+    std::ofstream ofs(file2);
+
+    if ((!ifs.is_open()) || (!ofs.is_open()))
+        return -1;
+    ofs << ifs.rdbuf();
+    return 0;
+}
+
 static int InitCgroup(int multiplier) {
     const char *cgroupPath = SrcCgroup();
 
     if (!cgroupPath)
         return -1;
-    std::string create_dir("mkdir " + std::string(cgroupPath) + "/butterfly");
-
-    BASH(create_dir) {
-        LOG_WARNING_("can't create butterfly cgroup, fail cmd '%s'",
-                     create_dir.c_str());
+    std::string directory = std::string(cgroupPath) + "/butterfly";
+    if (mkdir(directory.c_str(), S_IWGRP) < 0) {
+        LOG_WARNING_("can't create directory %s\n", directory.c_str());
     }
-    BASH("echo $(( `cat " + std::string(cgroupPath) + "/cpu.shares` * " +
-         std::to_string(multiplier) + " )) > " +
-         cgroupPath + "/butterfly/cpu.shares") {
+    std::string fileToRead = std::string(cgroupPath) + "/cpu.shares";
+    std::string fileToWrite = std::string(cgroupPath) + "/butterfly/cpu.shares";
+    std::ifstream in(fileToRead);
+    std::ofstream out(fileToWrite);
+    if ((in.is_open()) && (out.is_open())) {
+        int nb;
+        in >> nb;
+        std::string s = std::to_string(nb * multiplier);
+        out << s << std::endl;
+    } else {
         LOG_WARNING_("can't set cgroup priority");
     }
 
     // new cgroup use a diferent directory for cpushare and cpu
     // if not, we need to initilize some files
-    BASH(std::string("! cat ") + cgroupPath + "/cpuset.mems >> /dev/null") {
-            BASH("echo `cat " + std::string(cgroupPath) + "/cpuset.mems` > " +
-                 cgroupPath + "/butterfly/cpuset.mems") {
-                    LOG_WARNING_("can't set cgroup cpuset.mems");
-            }
-            BASH("echo `cat " + std::string(cgroupPath) + "/cpuset.cpus` > " +
-                 cgroupPath + "/butterfly/cpuset.cpus") {
-                    LOG_WARNING_("can't set cgroup cpuset.cpus");
-            }
+    fileToRead = std::string(cgroupPath) + "/cpuset.mems";
+    if (!access(fileToRead.c_str(), R_OK)) {
+        fileToWrite = std::string(cgroupPath) + "/butterfly/cpuset.mems";
+        if (setCgroups(fileToRead, fileToWrite) != 0) {
+            LOG_WARNING_("can't set cgroup cpuset.mems");
+        }
+
+        fileToRead = std::string(cgroupPath) + "/cpuset.cpus";
+        fileToWrite = std::string(cgroupPath) + "/butterfly/cpuset.mems";
```

Comment:
s/cpuset.mems/cpuset.cpus/

Reply:
this code was useful for centos6, so we might need to test it there before merging